added openocd image on top of arm image

kikass13 opened the pull request:
@rleh 

 * added an image on top of the existing arm-cortex image which downloads and installs openocd.

We can now create a docker container similiar to this

```
docker run -it --rm    \
   -v $PWD:$PWD --user $(id -u $USER) -e DIR=$PWD  \
   --device=/dev/bus/usb/003/007:/dev/bus/usb/003/007  \
   ghcr.io/modm-ext/modm-build-cortex-m-openocd:testing   \
   /bin/bash -c 'cd $DIR && ls && scons program'
```

whose job it is to mount a device (in this case a specific usb device), mount the current dir and call `scons program` to flash a connected target

kikass13 commented:
FYI:

I have utilized this in a python script, which simply runs a docker image via the docker api:

```python
def run(self, url, cmds, dir):
        result = False
        if self.client:
            ### prepare mounted devices if existing
            ### https://docker-py.readthedocs.io/en/stable/containers.html
            ### devices (list)
            ###   Expose host devices to the container, as a list of strings in the form <path_on_host>:<path_in_container>:<cgroup_permissions>.
            ###   For example, /dev/sda:/dev/xvda:rwm allows the container to have read-write access to the host’s /dev/sda via a node named /dev/xvda inside the container.
            deviceList = [os.path.join(dp, f) for dp, dn, filenames in os.walk("/dev/bus/usb/", followlinks=False) for f in filenames]
            devices = ["%s:%s:rmw" % (path, path) for path in deviceList]

            ### https://docker-py.readthedocs.io/en/stable/api.html#docker.types.Mount#
            ###   target (string) – Container path.
            ###   source (string) – Mount source (e.g. a volume name or a host path).
            ###   type (string) – The mount type (bind / volume / tmpfs / npipe). Default: volume.
            mountPoints = []
            if dir:
                mountPoints.append(docker.types.Mount(target=dir, source=dir, type="bind"))
            self.parent.log("Running '%s' with %s" % (url, cmds))
            for mp in mountPoints:
                self.parent.log("  - Mounting '%s'" % mp["Source"])

            ### https://docker-py.readthedocs.io/en/stable/containers.html
            try:
                container = self.client.containers.run(url, cmds, detach=True, devices=devices, mounts=mountPoints, stream=True, user=os.getuid())
            except docker.errors.ContainerError as e:
                self.parent.error("%s [%s]\nStatus: %s\nCommand: %s\nError: %s" % (e.container, e.image, e.exit_status, e.command, e.stderr.decode("utf-8").strip()))
            except docker.errors.APIError as ee:
                self.parent.error("Docker API Error")
                self.parent.error(ee)

            if container:
                try:
                   for log in container.logs(stdout=True, stderr=True, stream=True):
                       print(log.decode("utf-8").strip())
                except docker.errors.NotFound as ee:
                   pass
                result=True
                ### remove container
                container.remove(force=True)
        else:
            self.parent.error("Image '%s' could not be run with '%s'" % (url, ", ".join(cmds)))
        return result
```

rleh commented on Makefile:
```diff
@@ -1,4 +1,4 @@
-TARGET_IMAGES = avr cortex-m risc-v
+TARGET_IMAGES = cortex-m-openocd
```
Nope, just add your new build target to the list.

```suggestion
TARGET_IMAGES = avr cortex-m risc-v cortex-m-openocd
```

kikass13 replied:
wups, i removed those for testing and diddnt add trhem back :dagger: 
my bad

rleh commented on Makefile:
```diff
@@ -25,6 +25,10 @@ cortex-m: cortex-m.Dockerfile
 	cat $< | sed "s/modm-build-base:TAG/modm-build-base:$(DATE)/g" | \
 	docker build --tag ghcr.io/modm-ext/modm-build-$@:$(DATE) -
 	docker tag ghcr.io/modm-ext/modm-build-$@:$(DATE) ghcr.io/modm-ext/modm-build-$@:testing
+cortex-m-openocd: cortex-m-openocd.Dockerfile
```
Your new image depends on the `cortex-m` image for building:

```suggestion
cortex-m-openocd: cortex-m-openocd.Dockerfile cortex-m
```

rleh commented on cortex-m-openocd.Dockerfile:
```diff
@@ -0,0 +1,8 @@
+FROM ghcr.io/modm-ext/modm-build-base:TAG
+LABEL maintainer="Niklas Hauser <niklas.hauser@rwth-aachen.de>, Raphael Lehmann <raphael+docker@rleh.de>, Christopher Durand <christopher.durand@rwth-aachen.de>"
```
Feel free to add yourself as a maintainer!

kikass13 replied:
do I really have to? :D

rleh replied:
We will forward problems and support for the OpenOCD stuff to you anyway. 😉

rleh reviewed (changes requested):
Thanks, very nice!

I have a few small comments and change requests:
- Please update the README, preferably including your Docker command as an example.
  ```bash
  docker run -it --rm -v $PWD:$PWD --user $(id -u $USER) -e DIR=$PWD --device=/dev/bus/usb/003/007:/dev/bus/usb/003/007 ghcr.io/modm-ext/modm-build-cortex-m-openocd:latest /bin/bash -c 'cd $DIR && ls && scons program'
  ```
- (see below)

rleh commented:
On, and you also have to add the new image to the CI job: https://github.com/modm-ext/docker-modm-build/blob/master/.github/workflows/docker-build.yml

kikass13 commented:
* I added your fixes / suggestions
* I have added the docker run command to the readme (inside a new section) 
* I have added the image to the action

chris-durand commented:
Could you squash the last commit (the "Merge branch" one)?

kikass13 committed:
Add cortex-m-openocd image based on cortex-m

Co-authored-by: Raphael Lehmann <raphael@rleh.de>

kikass13 commented:
@chris-durand 
squashed to only 1 commit :)

chris-durand commented:
> squashed to only 1 commit :)

Thanks :)

rleh commented:
@kikass13 You also have to extend the docker image "TAG" preprocessing for the CI in [L21-L25](https://github.com/modm-ext/docker-modm-build/blob/master/.github/workflows/docker-build.yml#L21-L25)

e.g. by adding the following line after L25:
```
sed -i "s/modm-build-cortex-m:TAG/modm-build-cortex-m:${{ env.VERSION }}/g" cortex-m-openocd.Dockerfile
```

rleh commented on README.md:
````diff
@@ -15,3 +15,13 @@ make all -j4
 make push
 ```
+
+## Custom Images
````
```suggestion
### Image `modm-ext/modm-build-cortex-m-openocd` with OpenOCD
```

... and move it before the _Build images_ section!

rleh commented on README.md:
````diff
@@ -15,3 +15,13 @@ make all -j4
+
+You can program ARM Cortex-M (STM32, SAMD, ...) targets using the `cortex-m-openocd.Dockerfile` image like this:
+
+* ```
+  docker run -it --rm -v $PWD:$PWD --user $(id -u $USER) -e DIR=$PWD --device=/dev/bus/usb/003/007:/dev/bus/usb/003/007 ghcr.io/modm-ext/modm-build-cortex-m-openocd:latest /bin/bash -c 'cd $DIR && ls && scons program'
+  ```
````
The enumeration looks weird.
```suggestion
\`\`\`
docker run -it --rm -v $PWD:$PWD --user $(id -u $USER) -e DIR=$PWD --device=/dev/bus/usb/003/007:/dev/bus/usb/003/007 ghcr.io/modm-ext/modm-build-cortex-m-openocd:latest /bin/bash -c 'cd $DIR && ls && scons program'
```

rleh commented on README.md:
````diff
@@ -15,3 +15,13 @@ make all -j4
+  docker run -it --rm -v $PWD:$PWD --user $(id -u $USER) -e DIR=$PWD --device=/dev/bus/usb/003/007:/dev/bus/usb/003/007 ghcr.io/modm-ext/modm-build-cortex-m-openocd:latest /bin/bash -c 'cd $DIR && ls && scons program'
+  ```
+
+  * where `$PWD`  is your modm project directory and `/dev/bus/usb/003/007:/dev/bus/usb/003/007` is the device, where your programmer is located (for example `ST-LINK/V2.1`)
````
```suggestion
where `$PWD`  is your modm project directory and `/dev/bus/usb/003/007:/dev/bus/usb/003/007` is the device, where your programmer is located (for example `ST-LINK/V2.1`)
```

kikass13 commented:
Added suggested changes from @rleh

rleh reviewed (approved):
Merging tomorrow!

rleh commented:
Builds locally on my machine, let's see if the CI also works.

rleh commented:
New image is now publicly available: https://github.com/modm-ext/docker-modm-build/pkgs/container/modm-build-cortex-m-openocd
